app/bot/broadcast_status.go

```diff
@@ -0,0 +1,115 @@
+package bot
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	log "github.com/go-pkgz/lgr"
+)
+
+const (
+	MsgBroadcastStarted  = "Вещание началось"
+	MsgBroadcastFinished = "Вещание завершилось"
+)
+
+// BroadcastStatus bot returns current broadcast status
+type BroadcastStatus struct {
+	// Params
+	broadcastUrl string        // Url for "ping"
+	pingInterval time.Duration // Ping interval
+	delayToOff   time.Duration // After this interval of not OK, status will be switcher to OFF
+
+	// State
+	status         bool  // current broadcast status
+	lastSentStatus *bool // by default this value is null to react on first message
+	offPeriod      time.Duration
+	lastCheck      time.Time
+
+	// ping func (for tests)
+	ping func(ctx context.Context, url string) bool
+}
+
+// NewBroadcastStatus starts status checking goroutine and returns bot instance
+func NewBroadcastStatus(ctx context.Context, broadcastUrl string, pingInterval time.Duration, delayToOff time.Duration) *BroadcastStatus {
+	log.Printf("[INFO] BroadcastStatus bot with %v", broadcastUrl)
+	b := &BroadcastStatus{broadcastUrl: broadcastUrl, pingInterval: pingInterval, delayToOff: delayToOff, lastCheck: time.Now(), ping: ping}
+	go b.checker(ctx)
+	return b
+}
+
+// OnMessage returns current broadcast status
+func (b *BroadcastStatus) OnMessage(_ Message) (response string, answer bool) {
+	if b.lastSentStatus != nil && b.status == *b.lastSentStatus {
+		return
+	}
+
+	answer = true
+	response = MsgBroadcastStarted
+
+	status := b.status
+	b.lastSentStatus = &status
+	if !b.status {
+		response = MsgBroadcastFinished
+	}
+	return
+}
+
+func (b *BroadcastStatus) checker(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(b.pingInterval):
+			b.check(ctx)
+		}
+	}
+}
+
+func (b *BroadcastStatus) check(ctx context.Context) {
+	defer func() {
+		b.lastCheck = time.Now()
```

Comment:
lastCheck seems to be local for the goroutine. Having it as a member/field exposed directly is kind of risky as someone may try to access it from the main goroutine (i.e. OnMessage). My suggestion is to move the declaration to the checker and pass it to the check call. I would even go one step further and combine checker and check into one. If you don't line body of the code in case, you could use local function.

Reply:
Fixed. But i leave boarth checker and chek funcs for better testing.